File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

def dfs(visited, point, graph):
    visited[point] = True
    for node in graph[point]:
        if visited[node] is False:
            dfs(visited, node, graph)

n, m = map(int, input().split())

graph = [[] for i in range(n+1)]

for i in range(m):
    u, v = map(int, input().split())

    graph[u].append(v)
    graph[v].append(u)

visited = [False] * (n+1)
cnt = 0
for i in range(1, n+1):
    if visited[i] is False:
        dfs(visited, i, graph)
        cnt += 1
        
print(cnt)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <stack>
#include <map>
#include <set>
using namespace std;

void DFS(map<int, stack<int>> &graph, vector<bool> &visited, int node){
    visited[node]=true;
    while(!graph[node].empty()){
        int next_node = graph[node].top();
        graph[node].pop();
        if(!visited[next_node])
            DFS(graph, visited, next_node);
    }
    return;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    cin >> n >> m;

    map<int, stack<int>> graph;
    set<int> node_s;
    for(int i=0; i<m; i++){
        int a, b;
        cin >> a >> b;

        graph[a].push(b);
        graph[b].push(a);
        node_s.insert(a);
        node_s.insert(b);
    }
    
    vector<bool> visited (n+1, false);
    int ans = 0;
    for(auto node = node_s.begin(); node != node_s.end(); node++)
        if(!visited[*node]){
            DFS(graph, visited, *node);
            ans++;
        }

    for(int i=1; i<n+1; i++){
        if(visited[i] == false)
            ans++;
    }
    cout << ans << '\n';
}